File: app/src/main/java/cn/ifafu/ifafu/mvp/syllabus/SyllabusActivity.java
package cn.ifafu.ifafu.mvp.syllabus;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.viewpager2.widget.ViewPager2;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.gyf.immersionbar.ImmersionBar;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.OnClick;
import cn.ifafu.ifafu.R;
import cn.ifafu.ifafu.app.Constant;
import cn.ifafu.ifafu.data.entity.Course;
import cn.ifafu.ifafu.data.entity.SyllabusSetting;
import cn.ifafu.ifafu.mvp.base.BaseActivity;
import cn.ifafu.ifafu.mvp.main.MainActivity;
import cn.ifafu.ifafu.mvp.syllabus_item.SyllabusItemActivity;
import cn.ifafu.ifafu.mvp.syllabus_setting.SyllabusSettingActivity;
import cn.ifafu.ifafu.util.ChineseNumbers;
import cn.ifafu.ifafu.view.adapter.SyllabusPageAdapter;
import cn.ifafu.ifafu.view.dialog.ProgressDialog;

public class SyllabusActivity extends BaseActivity<SyllabusContract.Presenter>
        implements SyllabusContract.View, View.OnLongClickListener {

    @BindView(R.id.iv_background)
    ImageView ivBackground;
    @BindView(R.id.view_pager)
    ViewPager2 viewPager;
    @BindView(R.id.tv_subtitle)
    TextView tvSubtitle;
    @BindView(R.id.tb_syllabus)
    RelativeLayout tbSyllabus;

    private SyllabusPageAdapter mPageAdapter;

    private int mCurrentWeek = 1;

    private ProgressDialog progressDialog;

    public static final int BUTTON_ADD = 0;

    @Override
    public int initLayout(@Nullable Bundle savedInstanceState) {
        return R.layout.activity_syllabus;
    }

    @Override
    public void initData(@Nullable Bundle savedInstanceState) {
        ImmersionBar.with(this)
                .titleBarMarginTop(tbSyllabus)
                .statusBarDarkFont(true)
                .init();

        mPresenter = new SyllabusPresenter(this);

        progressDialog = new ProgressDialog(this);
        progressDialog.setText("加载中");
        TextView dateTV = findViewById(R.id.tv_date);
        dateTV.setText(new SimpleDateFormat("MM月dd日", Locale.CHINA).format(new Date()));

        viewPager = findViewById(R.id.view_pager);
        tvSubtitle.setOnLongClickListener(this);
        viewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                String numCN = ChineseNumbers.englishNumberToChinese(String.valueOf(position + 1));
                if (mCurrentWeek - 1 < 0) {
                    if (position != 0) {
                        tvSubtitle.setText(getString(R.string.week_format_not_return, numCN, "第一周"));
                    } else {
                        tvSubtitle.setText(getString(R.string.week_format_not, numCN));
                    }
                } else if (position != mCurrentWeek - 1) {
                    if (mCurrentWeek < 0) {
                        tvSubtitle.setText(getString(R.string.week_format_not_return, numCN, "第一周"));
                    } else {
                        tvSubtitle.setText(getString(R.string.week_format_not_return, numCN, "本周"));
                    }
                } else {
                    tvSubtitle.setText(getString(R.string.week_format, numCN));
                }
            }
        });
    }

    @Override
    public void setSyllabusSetting(SyllabusSetting setting) {
            Glide.with(this)
                    .load(setting.getBackground())
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(ivBackground);
        if (mPageAdapter == null) {
            mPageAdapter = new SyllabusPageAdapter(this, setting);
            mPageAdapter.setCourserClickListener((v, course) ->{
                Intent intent = new Intent(this, SyllabusItemActivity.class);
                intent.putExtra("course_id", ((Course) course.getOther()).getId());
                startActivityForResult(intent, Constant.SYLLABUS_ITEM_ACTIVITY);
            });
            viewPager.setAdapter(mPageAdapter);
        } else {
            mPageAdapter.setSyllabusSetting(setting);
            mPageAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void setSyllabusDate(List<Course> courses) {
        mPageAdapter.setCourseList(courses);
    }

    @Override
    public void redrawSyllabus() {
        mPageAdapter.notifyDataSetChanged();
    }

    @Override
    public void showLoading() {
        progressDialog.show();
    }

    @Override
    public void hideLoading() {
        progressDialog.cancel();
    }

    @OnClick({R.id.btn_add, R.id.btn_refresh, R.id.btn_back, R.id.btn_setting})
    public void onViewClicked(View v) {
        switch (v.getId()) {
            case R.id.btn_add:
                Intent intent = new Intent(this, SyllabusItemActivity.class);
                intent.putExtra("come_from", BUTTON_ADD);
                startActivityForResult(intent, Constant.SYLLABUS_ITEM_ACTIVITY);
                break;
            case R.id.btn_refresh:
                mPresenter.updateSyllabusNet();
                break;
            case R.id.btn_back:
                onFinishActivity();
                break;
            case R.id.btn_setting:
                startActivityForResult(new Intent(this, SyllabusSettingActivity.class),
                        Constant.SYLLABUS_SETTING_ACTIVITY);
                break;
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_DOWN) {
            onFinishActivity();
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    private void onFinishActivity() {
        switch (getIntent().getIntExtra("from", -1)) {
            case Constant.SYLLABUS_WIDGET:
            case Constant.SPLASH_ACTIVITY:
                Intent intent = new Intent(this, MainActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
                startActivity(intent);
                break;
        }
        finish();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == Constant.SYLLABUS_ITEM_ACTIVITY && resultCode == Activity.RESULT_OK) {
            mPresenter.updateSyllabusLocal();
            return;
        } else if (requestCode == Constant.SYLLABUS_SETTING_ACTIVITY) {
            mPresenter.updateSyllabusSetting();
            return;
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public boolean onLongClick(View v) {
        if (v.getId() == R.id.tv_subtitle) {
            viewPager.setCurrentItem(mCurrentWeek - 1, true);
            return true;
        }
        return false;
    }
}

File: app/src/main/java/cn/ifafu/ifafu/mvp/syllabus/SyllabusActivity.kt
package cn.ifafu.ifafu.mvp.syllabus

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.View
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import cn.ifafu.ifafu.R
import cn.ifafu.ifafu.app.Constant
import cn.ifafu.ifafu.base.mvp.BaseActivity
import cn.ifafu.ifafu.entity.Course
import cn.ifafu.ifafu.entity.SyllabusSetting
import cn.ifafu.ifafu.mvp.main.MainActivity
import cn.ifafu.ifafu.mvp.syllabus_item.SyllabusItemActivity
import cn.ifafu.ifafu.mvp.syllabus_setting.SyllabusSettingActivity
import cn.ifafu.ifafu.util.ChineseNumbers
import cn.ifafu.ifafu.view.adapter.SyllabusPageAdapter
import cn.ifafu.ifafu.view.dialog.LoadingDialog
import cn.ifafu.ifafu.view.syllabus.CourseBase
import com.gyf.immersionbar.ImmersionBar
import kotlinx.android.synthetic.main.syllabus_activity.*
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList

class SyllabusActivity : BaseActivity<SyllabusContract.Presenter>(), SyllabusContract.View,
        View.OnClickListener, View.OnLongClickListener {

    private var mCurrentWeek = 1
    private lateinit var mPageAdapter: SyllabusPageAdapter
    private lateinit var loadingDialog: LoadingDialog

    override fun getLayoutId(savedInstanceState: Bundle?): Int {
        return R.layout.syllabus_activity
    }

    override fun initData(savedInstanceState: Bundle?) {
        ImmersionBar.with(this)
                .titleBarMarginTop(tb_syllabus)
                .statusBarDarkFont(true)
                .init()

        mPresenter = SyllabusPresenter(this)

        loadingDialog = LoadingDialog(this)
        loadingDialog.setText("加载中")
        loadingDialog.setOnCancelListener {
            mPresenter?.cancelLoading()
        }

        btn_back.setOnClickListener(this)
        btn_add.setOnClickListener(this)
        btn_refresh.setOnClickListener(this)
        btn_setting.setOnClickListener(this)

        tv_date.text = SimpleDateFormat("MM月dd日", Locale.CHINA).format(Date())
        tv_subtitle.setOnLongClickListener(this)
    }

    override fun setSyllabusSetting(setting: SyllabusSetting) {
        val themeColor = setting.themeColor
        tv_date.setTextColor(themeColor)
        tv_subtitle.setTextColor(themeColor)
        btn_back.setColorFilter(themeColor)
        btn_add.setColorFilter(themeColor)
        btn_refresh.setColorFilter(themeColor)
        btn_setting.setColorFilter(themeColor)
        ImmersionBar.with(this)
                .statusBarDarkFont(setting.statusDartFont)
                .init()

        mPageAdapter = SyllabusPageAdapter(setting, ArrayList()) {
            val intent = Intent(this, SyllabusItemActivity::class.java)
            intent.putExtra("course_id", (it.getOther() as Course).id)
            startActivityForResult(intent, Constant.ACTIVITY_SYLLABUS_ITEM)
        }
        view_pager.adapter = mPageAdapter
        setCurrentWeek(setting.currentWeek)
        Log.d(TAG, "background: ${setting.background}")

        if (setting.background.isNotEmpty()) {
//            Glide.with(this)
//                    .load(setting.background)
//                    .skipMemoryCache(true)
//                    .transition(DrawableTransitionOptions.withCrossFade())
//                    .into(iv_background)
            iv_background.setImageDrawable(null)
            iv_background.setImageURI(Uri.parse(setting.background))
        } else {
            iv_background.setImageDrawable(null)
        }
    }

    override fun setCurrentWeek(currentWeek: Int) {
        mCurrentWeek = currentWeek
        view_pager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                val numCN: String? = ChineseNumbers.englishNumberToChinese((position + 1).toString())
                if (mCurrentWeek - 1 < 0) {
                    if (position != 0) {
                        tv_subtitle.text = getString(R.string.week_format_not_return, numCN, "第一周")
                    } else {
                        tv_subtitle.text = getString(R.string.week_format_not, numCN)
                    }
                } else if (position != mCurrentWeek - 1) {
                    if (mCurrentWeek < 0) {
                        tv_subtitle.text = getString(R.string.week_format_not_return, numCN, "第一周")
                    } else {
                        tv_subtitle.text = getString(R.string.week_format_not_return, numCN, "本周")
                    }
                } else {
                    tv_subtitle.text = getString(R.string.week_format, numCN)
                }
            }
        })
        view_pager.setCurrentItem(currentWeek - 1, false)
    }

    override fun setSyllabusData(courses: MutableList<MutableList<CourseBase>?>) {
        mPageAdapter.courses = courses
    }

    override fun showLoading() {
        loadingDialog.show()
    }

    override fun hideLoading() {
        loadingDialog.cancel()
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.btn_add -> {
                val intent = Intent(this, SyllabusItemActivity::class.java)
                intent.putExtra("come_from", BUTTON_ADD)
                startActivityForResult(intent, Constant.ACTIVITY_SYLLABUS_ITEM)
            }
            R.id.btn_refresh -> mPresenter.updateSyllabusNet()
            R.id.btn_back -> onFinishActivity()
            R.id.btn_setting -> startActivityForResult(
                    Intent(this, SyllabusSettingActivity::class.java),
                    Constant.ACTIVITY_SYLLABUS_SETTING)
        }
    }

    override fun onLongClick(v: View?): Boolean {
        return when (v?.id) {
            R.id.tv_subtitle -> {
                view_pager.setCurrentItem(mCurrentWeek - 1, true)
                true
            }
            else -> false
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK && event?.action == KeyEvent.ACTION_DOWN) {
            onFinishActivity()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    private fun onFinishActivity() {
        when (intent.getIntExtra("from", -1)) {
            Constant.SYLLABUS_WIDGET, Constant.ACTIVITY_SPLASH -> {
                val intent = Intent(this, MainActivity::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
                startActivity(intent)
            }
        }
        finish()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == Constant.ACTIVITY_SYLLABUS_ITEM && resultCode == Activity.RESULT_OK) {
            mPresenter.updateSyllabusLocal()
        } else if (requestCode == Constant.ACTIVITY_SYLLABUS_SETTING && resultCode == Activity.RESULT_OK) {
            mPresenter.updateSyllabusSetting()
            mPresenter.updateSyllabusLocal()
        } else {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    companion object {
        const val BUTTON_ADD = 0
    }
}